tools/crates/cli/src/contexts/app.rs

```diff
@@ -0,0 +1,182 @@
+use crate::commands::subcommand::{AllSubCommands, RunAllSubCommand};
+
+#[derive(HasField)]
+pub struct ToolApp {
+    pub config_path: PathBuf,
+    pub runtime: HermesRuntime,
+}
+
+pub struct ToolAppComponents;
+
+pub struct ToolParserComponents;
+
+pub struct ToolCommandRunnerComponents;
+
+pub struct UpdateToolConfig;
+
+impl HasComponents for ToolApp {
+    type Components = ToolAppComponents;
+}
+
+delegate_components! {
+    ToolAppComponents {
+        [
+            ErrorTypeComponent,
+            ErrorRaiserComponent,
+            RetryableErrorComponent,
+        ]:
+            ProvideCliError,
+        RuntimeTypeComponent: WithType<HermesRuntime>,
+        RuntimeGetterComponent: WithField<symbol!("runtime")>,
+        [
+            LoggerTypeComponent,
+            LoggerGetterComponent,
+            GlobalLoggerGetterComponent,
+        ]:
+            ProvideHermesLogger,
+        ConfigTypeComponent:
+            WithType<StarknetRelayerConfig>,
+        BootstrapTypeComponent:
+            WithType<StarknetBootstrap>,
+        OutputTypeComponent:
+            WithType<()>,
+        BootstrapLoaderComponent:
+            LoadStarknetBootstrap,
+        ConfigPathGetterComponent:
+            GetDefaultConfigField,
+        ConfigLoaderComponent:
+            LoadTomlConfig,
+        ConfigWriterComponent:
+            WriteTomlConfig,
+        CommandRunnerComponent:
+            UseDelegate<ToolCommandRunnerComponents>,
+        BuilderLoaderComponent:
+            LoadStarknetBuilder,
+        BuilderTypeComponent:
+            WithType<StarknetBuilder>,
+        ArgParserComponent:
+            UseDelegate<ToolParserComponents>,
+    }
+}
+
+delegate_components! {
+    ToolCommandRunnerComponents {
+        AllSubCommands: RunAllSubCommand,
+
+        StarknetSubCommand: RunStarknetSubCommand,
+        TransferArgs: RunTransferArgs,
+    }
+}
+
+impl<App> ProvideAnyCounterparty<App> for ToolAppComponents
+where
+    App: Async,
+{
+    type AnyCounterparty = CosmosChain;
+}
+
+impl<Value> OutputProducer<ToolApp, Value> for ToolAppComponents {
+    fn produce_output(_app: &ToolApp, _value: Value) {}
+}
+
+impl ConfigUpdater<StarknetChainDriver, StarknetRelayerConfig> for UpdateToolConfig {
+    fn update_config(
+        chain_driver: &StarknetChainDriver,
+        config: &mut StarknetRelayerConfig,
+    ) -> Result<String, HermesError> {
+        if config.starknet_chain_config.is_some() {
+            return Err(StarknetChainDriver::raise_error(
+                "starknet chain config is already present in config file",
+            ));
+        }
+
+        let relayer_wallet = chain_driver
+            .wallets
+            .get("relayer")
+            .ok_or_else(|| StarknetChainDriver::raise_error("expect relayer wallet to be present"))?
+            .clone();
+
+        let chain_config = StarknetChainConfig {
+            json_rpc_url: format!("http://localhost:{}/", chain_driver.node_config.rpc_port),
```

Comment:
This should be [SocketAddr](https://doc.rust-lang.org/std/net/enum.SocketAddr.html). The public endpoint will have an IP.

Reply:
Done in 3a61fb1907611e90e9923218b230f6638df1fb49

Reply:
Will add the configurable IP in another commit

Reply:
Added a `rpc_addr` field to the `StarknetNodeConfig` but the bootstrap currently uses localhost and a random free port. See https://github.com/informalsystems/ibc-starknet/pull/232/files#diff-e7746581e0a2ba08412a5d6307cbe2118ef587b5ee2258a5b37534ed0df0c531R62

Reply:
This should have been `String` or `Url` because of #289. You were right. Sorry for my oversight.